Week9

Amnesiac9 opened the pull request.

Amnesiac9 committed:
Tables Added and First iteration of required procedures

Amnesiac9 committed:
Fixed Add Post procedure and New User Trigger

Amnesiac9 committed:
Fixed user_added trigger again

Amnesiac9 committed:
Fixed users getting notifications for their own joining

Amnesiac9 committed:
Fix incorrect name attributed to posts in notifications

Amnesiac9 committed:
Attempting to fix Notifications_VIEW

Amnesiac9 committed:
Attempting to fix notification_posts view

Amnesiac9 committed:
Added Logout Event

Amnesiac9 committed:
Attempt to fix logout event

Amnesiac9 committed:
Fixed Logout Event

Amnesiac9 committed:
Removed unnecessary comments

Amnesiac9 commented:
Github does weird things to my tab formatting, it looks fine and then once I commit the changes it's all messed up.

Amnesiac9 committed:
Formatting

halverneus commented on week9/answers.sql:
```diff
@@ -1,3 +1,195 @@
+        
+-- ------------------ Notify All Procedure ------------------ 
+
+DELIMITER ;;
```
You can group all of your procedure declarations within a single DELIMITER scope. Aka:

```sql
DELIMITER ;;
CREATE PROCEDURE blah...
END;;
CREATE PROCEDURE halb...
END;;
CREATE TRIGGER ...
...
DELIMITER ;
```

Amnesiac9 replied:
Ah good tip

halverneus commented on week9/answers.sql:
```diff
@@ -1,3 +1,195 @@
+    INSERT INTO posts
+		(user_id, content)
+	VALUES
+		(NEW.user_id, CONCAT(first_name_new, ' ', last_name_new, ' ', 'just joined!'));
```
You can save yourself some code and SELECT statements by just using the first and last name supplied by the trigger:

```sql
INSERT INTO posts
    (user_id, content)
VALUES
    (NEW.user_id, CONCAT(NEW.first_name, ' ', NEW.last_name, ' just joined!'));
```

Amnesiac9 replied:
Fixed! I swear i tried that and it didn't work... I must have typed something wrong! lol

halverneus commented on week9/answers.sql:
```diff
@@ -1,3 +1,195 @@
+	VALUES
+		(NEW.user_id, CONCAT(first_name_new, ' ', last_name_new, ' ', 'just joined!'));
+        
+	SELECT LAST_INSERT_ID() FROM posts LIMIT 1 INTO this_post_id;
```
`LIMIT 1` is unnecessary, here. `LAST_INSERT_ID()` can only every return a single value.

Amnesiac9 replied:
It's giving me this error when I don't have LIMIT 1:

Error Code: 1172. Result consisted of more than one row	0.000 sec

halverneus replied:
Try simply:

```sql
SELECT LAST_INSERT_ID() INTO this_post_id;
```

LAST_INSERT_ID() operates on a connection, not based on a selected table (so, if you happened to do your last insert in the `users` table and ran this query, it would still only give you the ID from the `users` table despite the `SELECT ... FROM posts`.

halverneus commented on week9/answers.sql:
```diff
@@ -1,3 +1,195 @@
+    SET cur_user = 1;
+	
+    -- Loop through all users and add notification for them
+    WHILE cur_user < user_count DO
```
Clever solution, but if a user gets deleted (which happens in the `checks.yml`, though we don't see this bug because a user is never added, afterwards (a use-case for next year)), you'll be attempting to add a notification for a `cur_user` that doesn't exist and will get a foreign key exception (with no logic to catch the error, this function will be perpetually broken (especially if the first user deletes their account)). Does that make sense? I recommend using a `CURSOR` so you can be certain you are adding notifications for users that exist.

Amnesiac9 replied:
Ah didn't think about the possibility of a user ID not existing. Cursor seems the best solution, added!

halverneus commented on week9/answers.sql:
```diff
@@ -1,3 +1,195 @@
+    INSERT INTO posts (user_id, content) VALUES (this_user_id, this_content);
+
+    -- Get current post ID
+    SELECT LAST_INSERT_ID() FROM posts LIMIT 1 INTO this_post_id;
```
`LIMIT 1` not required. `LAST_INSERT_ID()` can only ever return a single value.

Amnesiac9 replied:
It's giving me this error when I don't have LIMIT 1:

Error Code: 1172. Result consisted of more than one row 0.000 sec

halverneus replied:
Try simply:

```sql
SELECT LAST_INSERT_ID() INTO this_post_id;
```

LAST_INSERT_ID() operates on a connection, not based on a selected table (so, if you happened to do your last insert in the `users` table and ran this query, it would still only give you the ID from the `users` table despite the `SELECT ... FROM posts`.

Amnesiac9 committed:
Added suggested improvmenets

-Used cursor instead of loop in notify all users procedure.
-Removed repeated DELIMETER changes.
-Used NEW. instead of creating and selecting into variables for new user notification.

Amnesiac9 committed:
Fixed users getting notifications for their own account creation

Amnesiac9 committed:
Attempt to fix missing notifications. Cursors are weird.

Amnesiac9 committed:
Attempt to fix users getting their own joining notification

Amnesiac9 committed:
Formatting and Comments

Amnesiac9 committed:
Updated view to use Left Outer Joins

Amnesiac9 committed:
Fixed incorrect LAST_INSERT_ID() Usage
